File: api/reception.py
from flask import Blueprint, request, jsonify, g
from .extensions import supabase
import uuid
from datetime import datetime

reception_bp = Blueprint("reception", __name__)

# Middleware to check if module is enabled
def check_module_access():
    if not g.tenant_id:
        return jsonify({"error": "Tenant not found"}), 404
        
    if "reception" not in g.modules:
        return jsonify({"error": "Reception module is not enabled for this tenant"}), 403
        
    return None

# Get today's appointments
@reception_bp.route("/reception/appointments/today", methods=["GET"])
def get_today_appointments():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # Get today's date
        today = datetime.now().strftime("%Y-%m-%d")
        
        # Get appointments for today
        result = supabase.table("appointments").select("*").eq("client_id", g.tenant_id).eq("date", today).execute()
        
        return jsonify(result.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Register patient
@reception_bp.route("/reception/patients", methods=["POST"])
def register_patient():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        patient_data = request.json
        
        # Generate patient ID
        patient_id = str(uuid.uuid4())
        
        # Prepare patient data
        new_patient = {
            "id": patient_id,
            "client_id": g.tenant_id,
            "registered_at": datetime.now().isoformat(),
            **patient_data
        }
        
        # Insert patient
        result = supabase.table("patients").insert(new_patient).execute()
        
        return jsonify(result.data[0]), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Book appointment
@reception_bp.route("/reception/appointments", methods=["POST"])
def book_appointment():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        appointment_data = request.json
        
        # Generate appointment ID
        appointment_id = str(uuid.uuid4())
        
        # Get doctor name
        doctor_id = appointment_data.get("doctorId")
        doctor_query = supabase.table("staff").select("name").eq("id", doctor_id).single().execute()
        doctor_name = doctor_query.data["name"] if doctor_query.data else "Unknown Doctor"
        
        # Prepare appointment data
        new_appointment = {
            "id": appointment_id,
            "client_id": g.tenant_id,
            "doctor_name": doctor_name,
            "status": "confirmed",
            "booked_at": datetime.now().isoformat(),
            **appointment_data
        }
        
        # Insert appointment
        result = supabase.table("appointments").insert(new_appointment).execute()
        
        return jsonify(result.data[0]), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get queue list
@reception_bp.route("/reception/queue", methods=["GET"])
def get_queue_list():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        result = supabase.table("queue").select("*").eq("client_id", g.tenant_id).execute()
        
        return jsonify(result.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Update patient status in queue
@reception_bp.route("/reception/queue/<patient_id>/status", methods=["PATCH"])
def update_patient_status(patient_id):
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        new_status = request.json.get("status")
        
        if not new_status:
            return jsonify({"error": "Status is required"}), 400
            
        # Update patient status
        result = supabase.table("queue").update({
            "status": new_status
        }).eq("id", patient_id).eq("client_id", g.tenant_id).execute()
        
        if not result.data:
            return jsonify({"error": "Patient not found in queue"}), 404
            
        # Get updated queue
        queue = supabase.table("queue").select("*").eq("client_id", g.tenant_id).execute()
        
        return jsonify(queue.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Add patient to queue
@reception_bp.route("/reception/queue", methods=["POST"])
def add_to_queue():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        patient_data = request.json
        
        # Generate queue entry ID
        entry_id = str(uuid.uuid4())
        
        # Get current queue to determine queue number
        queue_query = supabase.table("queue").select("*").eq("client_id", g.tenant_id).execute()
        queue_number = len(queue_query.data) + 1
        
        # Prepare queue entry data
        queue_entry = {
            "id": entry_id,
            "client_id": g.tenant_id,
            "status": "waiting",
            "checked_in_at": datetime.now().isoformat(),
            "queue_number": queue_number,
            **patient_data
        }
        
        # Insert queue entry
        result = supabase.table("queue").insert(queue_entry).execute()
        
        return jsonify(result.data[0]), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Upload consent form
@reception_bp.route("/reception/consent", methods=["POST"])
def upload_consent():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # In a real app, this would handle file upload to storage
        # For this demo, we'll just create a record
        
        patient_id = request.form.get("patientId")
        patient_name = request.form.get("patientName")
        signature = request.form.get("signature")
        
        if not patient_id or not patient_name:
            return jsonify({"error": "Patient ID and name are required"}), 400
            
        # Generate consent form ID
        form_id = str(uuid.uuid4())
        
        # Mock file data
        file = request.files.get("file")
        file_type = file.content_type if file else "application/pdf"
        file_name = file.filename if file else "consent.pdf"
        
        # Prepare consent form data
        consent_form = {
            "id": form_id,
            "client_id": g.tenant_id,
            "patient_id": patient_id,
            "patient_name": patient_name,
            "file_type": file_type,
            "file_name": file_name,
            "signature": signature,
            "uploaded_at": datetime.now().isoformat()
        }
        
        # Insert consent form
        result = supabase.table("consent_forms").insert(consent_form).execute()
        
        return jsonify(result.data[0]), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get doctors
@reception_bp.route("/reception/doctors", methods=["GET"])
def get_doctors():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # Get doctors
        result = supabase.table("staff") \
               .select("id, name, specialization, available") \
               .eq("client_id", g.tenant_id) \
               .eq("role", "doctor") \
               .execute()
               
        return jsonify(result.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get available time slots
@reception_bp.route("/reception/time-slots", methods=["GET"])
def get_available_time_slots():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        date = request.args.get("date")
        doctor_id = request.args.get("doctorId")
        
        if not date or not doctor_id:
            return jsonify({"error": "Date and doctor ID are required"}), 400
            
        # Get booked slots
        booked_query = supabase.table("appointments") \
                     .select("time") \
                     .eq("client_id", g.tenant_id) \
                     .eq("date", date) \
                     .eq("doctor_id", int(doctor_id)) \
                     .execute()
                     
        booked_slots = [appointment["time"] for appointment in booked_query.data]
        
        # All possible time slots
        all_slots = [
            "09:00", "09:30", "10:00", "10:30", "11:00", "11:30",
            "12:00", "12:30", "14:00", "14:30", "15:00", "15:30",
            "16:00", "16:30", "17:00", "17:30"
        ]
        
        # Filter out booked slots
        available_slots = [slot for slot in all_slots if slot not in booked_slots]
        
        return jsonify(available_slots), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get reception stats
@reception_bp.route("/reception/stats", methods=["GET"])
def get_reception_stats():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # Get today's date
        today = datetime.now().strftime("%Y-%m-%d")
        
        # Get today's appointments
        appointments_query = supabase.table("appointments") \
                           .select("*") \
                           .eq("client_id", g.tenant_id) \
                           .eq("date", today) \
                           .execute()
                           
        # Get today's registrations
        registrations_query = supabase.table("patients") \
                            .select("*") \
                            .eq("client_id", g.tenant_id) \
                            .gte("registered_at", f"{today}T00:00:00") \
                            .execute()
                            
        # Get queue
        queue_query = supabase.table("queue") \
                    .select("*") \
                    .eq("client_id", g.tenant_id) \
                    .execute()
                    
        # Calculate stats
        today_appointments = len(appointments_query.data)
        walk_ins_registered = len(registrations_query.data)
        patients_in_queue = len(queue_query.data)
        completed_appointments = len([a for a in appointments_query.data if a["status"] == "completed"])
        
        stats = {
            "todayAppointments": today_appointments,
            "walkInsRegistered": walk_ins_registered,
            "patientsInQueue": patients_in_queue,
            "completedAppointments": completed_appointments
        }
        
        return jsonify(stats), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

File: src/api/reception.ts
// Mock API functions for Reception module
// In a real application, these would make HTTP requests to your backend

// Types
export interface Appointment {
  id: string;
  patientName: string;
  patientPhone: string;
  doctorId: number;
  doctorName: string;
  date: string;
  time: string;
  status: 'confirmed' | 'waiting' | 'in-progress' | 'completed' | 'cancelled';
  type: 'consultation' | 'follow-up' | 'emergency' | 'specialist';
  notes?: string;
  bookedAt?: string;
}

export interface Patient {
  id: string;
  fullName: string;
  mobile: string;
  email?: string;
  gender: 'male' | 'female' | 'other';
  dateOfBirth: string;
  appointmentType: 'new' | 'follow-up';
  referredBy: 'google' | 'instagram' | 'facebook' | 'referral' | 'walk-in' | 'other';
  clinicBranch?: string;
  registeredAt: string;
}

export interface Doctor {
  id: number;
  name: string;
  specialization: string;
  available: boolean;
}

export interface QueueEntry {
  id: string;
  patientName: string;
  patientPhone: string;
  doctorName?: string;
  appointmentTime?: string;
  status: 'waiting' | 'checked-in' | 'with-doctor' | 'completed' | 'cancelled';
  checkedInAt: string;
  queueNumber: number;
}

export interface ConsentForm {
  id: string;
  patientId: string;
  patientName: string;
  fileType: string;
  fileName: string;
  signature?: string;
  uploadedAt: string;
}

export interface ReceptionStats {
  todayAppointments: number;
  walkInsRegistered: number;
  patientsInQueue: number;
  completedAppointments: number;
}

// Mock data storage (using localStorage for persistence)
const STORAGE_KEYS = {
  APPOINTMENTS: 'hospverse_appointments',
  PATIENTS: 'hospverse_patients',
  QUEUE: 'hospverse_queue',
  CONSENT_FORMS: 'hospverse_consent_forms'
};

// Helper function to get data from localStorage
const getStorageData = <T>(key: string, defaultValue: T[] = []): T[] => {
  try {
    const data = localStorage.getItem(key);
    return data ? JSON.parse(data) : defaultValue;
  } catch (error) {
    console.error('Error reading from localStorage:', error);
    return defaultValue;
  }
};

// Helper function to save data to localStorage
const setStorageData = <T>(key: string, data: T[]): void => {
  try {
    localStorage.setItem(key, JSON.stringify(data));
  } catch (error) {
    console.error('Error saving to localStorage:', error);
  }
};

// Generate unique ID
const generateId = (): string => Date.now().toString(36) + Math.random().toString(36).substr(2);

// Mock doctors data
const DOCTORS: Doctor[] = [
  { id: 1, name: 'Dr. Sarah Johnson', specialization: 'General Medicine', available: true },
  { id: 2, name: 'Dr. Michael Chen', specialization: 'Dermatology', available: true },
  { id: 3, name: 'Dr. Emily Rodriguez', specialization: 'Cardiology', available: false },
  { id: 4, name: 'Dr. David Kim', specialization: 'Pediatrics', available: true },
  { id: 5, name: 'Dr. Lisa Thompson', specialization: 'Orthopedics', available: true }
];

// Mock time slots
const TIME_SLOTS: string[] = [
  '09:00', '09:30', '10:00', '10:30', '11:00', '11:30',
  '12:00', '12:30', '14:00', '14:30', '15:00', '15:30',
  '16:00', '16:30', '17:00', '17:30'
];

// Initialize with some mock data if localStorage is empty
const initializeMockData = (): void => {
  if (!localStorage.getItem(STORAGE_KEYS.APPOINTMENTS)) {
    const mockAppointments: Appointment[] = [
      {
        id: '1',
        patientName: 'John Doe',
        patientPhone: '+1-555-0123',
        doctorId: 1,
        doctorName: 'Dr. Sarah Johnson',
        date: new Date().toISOString().split('T')[0],
        time: '09:00',
        status: 'confirmed',
        type: 'consultation',
        notes: 'Regular checkup'
      },
      {
        id: '2',
        patientName: 'Jane Smith',
        patientPhone: '+1-555-0124',
        doctorId: 2,
        doctorName: 'Dr. Michael Chen',
        date: new Date().toISOString().split('T')[0],
        time: '10:30',
        status: 'waiting',
        type: 'follow-up',
        notes: 'Skin condition follow-up'
      }
    ];
    setStorageData(STORAGE_KEYS.APPOINTMENTS, mockAppointments);
  }

  if (!localStorage.getItem(STORAGE_KEYS.PATIENTS)) {
    const mockPatients: Patient[] = [
      {
        id: '1',
        fullName: 'John Doe',
        mobile: '+1-555-0123',
        email: 'john.doe@email.com',
        gender: 'male',
        dateOfBirth: '1990-05-15',
        appointmentType: 'new',
        referredBy: 'google',
        clinicBranch: 'main',
        registeredAt: new Date().toISOString()
      }
    ];
    setStorageData(STORAGE_KEYS.PATIENTS, mockPatients);
  }
};

// Initialize mock data
initializeMockData();

// API Functions
export const getTodayAppointments = async (): Promise<Appointment[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const appointments = getStorageData<Appointment>(STORAGE_KEYS.APPOINTMENTS);
      const today = new Date().toISOString().split('T')[0];
      const todayAppointments = appointments.filter(apt => apt.date === today);
      resolve(todayAppointments);
    }, 500);
  });
};

export const registerPatient = async (patientData: Omit<Patient, 'id' | 'registeredAt'>): Promise<Patient> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        const patients = getStorageData<Patient>(STORAGE_KEYS.PATIENTS);
        const newPatient: Patient = {
          id: generateId(),
          ...patientData,
          registeredAt: new Date().toISOString()
        };
        patients.push(newPatient);
        setStorageData(STORAGE_KEYS.PATIENTS, patients);
        resolve(newPatient);
      } catch (error) {
        reject(error);
      }
    }, 800);
  });
};

export const bookAppointment = async (appointmentData: Omit<Appointment, 'id' | 'doctorName' | 'status' | 'bookedAt'>): Promise<Appointment> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        const appointments = getStorageData<Appointment>(STORAGE_KEYS.APPOINTMENTS);
        const doctor = DOCTORS.find(d => d.id === appointmentData.doctorId);
        
        const newAppointment: Appointment = {
          id: generateId(),
          ...appointmentData,
          doctorName: doctor?.name || 'Unknown Doctor',
          status: 'confirmed',
          bookedAt: new Date().toISOString()
        };
        
        appointments.push(newAppointment);
        setStorageData(STORAGE_KEYS.APPOINTMENTS, appointments);
        resolve(newAppointment);
      } catch (error) {
        reject(error);
      }
    }, 800);
  });
};

export const getQueueList = async (): Promise<QueueEntry[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const queue = getStorageData<QueueEntry>(STORAGE_KEYS.QUEUE);
      resolve(queue);
    }, 300);
  });
};

export const updatePatientStatus = async (patientId: string, newStatus: QueueEntry['status']): Promise<QueueEntry[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const queue = getStorageData<QueueEntry>(STORAGE_KEYS.QUEUE);
      const updatedQueue = queue.map(patient => 
        patient.id === patientId ? { ...patient, status: newStatus } : patient
      );
      setStorageData(STORAGE_KEYS.QUEUE, updatedQueue);
      resolve(updatedQueue);
    }, 300);
  });
};

export const addToQueue = async (patientData: Omit<QueueEntry, 'id' | 'status' | 'checkedInAt' | 'queueNumber'>): Promise<QueueEntry> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const queue = getStorageData<QueueEntry>(STORAGE_KEYS.QUEUE);
      const queueEntry: QueueEntry = {
        id: generateId(),
        ...patientData,
        status: 'waiting',
        checkedInAt: new Date().toISOString(),
        queueNumber: queue.length + 1
      };
      queue.push(queueEntry);
      setStorageData(STORAGE_KEYS.QUEUE, [queueEntry]);
      resolve(queueEntry);
    }, 300);
  });
};

export const uploadConsent = async (formData: FormData): Promise<ConsentForm> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        const consentForms = getStorageData<ConsentForm>(STORAGE_KEYS.CONSENT_FORMS);
        const file = formData.get('file') as File;
        const newConsent: ConsentForm = {
          id: generateId(),
          patientId: formData.get('patientId') as string,
          patientName: formData.get('patientName') as string,
          fileType: file?.type || 'application/pdf',
          fileName: file?.name || 'consent.pdf',
          signature: formData.get('signature') as string,
          uploadedAt: new Date().toISOString()
        };
        consentForms.push(newConsent);
        setStorageData(STORAGE_KEYS.CONSENT_FORMS, consentForms);
        resolve(newConsent);
      } catch (error) {
        reject(error);
      }
    }, 1000);
  });
};

export const getDoctors = async (): Promise<Doctor[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(DOCTORS);
    }, 200);
  });
};

export const getAvailableTimeSlots = async (date: string, doctorId: string): Promise<string[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const appointments = getStorageData<Appointment>(STORAGE_KEYS.APPOINTMENTS);
      const bookedSlots = appointments
        .filter(apt => apt.date === date && apt.doctorId === parseInt(doctorId))
        .map(apt => apt.time);
      
      const availableSlots = TIME_SLOTS.filter(slot => !bookedSlots.includes(slot));
      resolve(availableSlots);
    }, 300);
  });
};

export const getReceptionStats = async (): Promise<ReceptionStats> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const appointments = getStorageData<Appointment>(STORAGE_KEYS.APPOINTMENTS);
      const patients = getStorageData<Patient>(STORAGE_KEYS.PATIENTS);
      const queue = getStorageData<QueueEntry>(STORAGE_KEYS.QUEUE);
      
      const today = new Date().toISOString().split('T')[0];
      const todayAppointments = appointments.filter(apt => apt.date === today);
      const todayRegistrations = patients.filter(patient => 
        patient.registeredAt.split('T')[0] === today
      );
      
      resolve({
        todayAppointments: todayAppointments.length,
        walkInsRegistered: todayRegistrations.length,
        patientsInQueue: queue.length,
        completedAppointments: todayAppointments.filter(apt => apt.status === 'completed').length
      });
    }, 400);
  });
};